src/libexpr/paths.cc

```diff
@@ -0,0 +1,92 @@
+#include "eval.hh"
+#include "util.hh"
+#include "fs-input-accessor.hh"
+
+namespace nix {
+
+SourcePath EvalState::rootPath(const Path & path)
+{
+    return {rootFS, CanonPath(path)};
+}
+
+void EvalState::registerAccessor(ref<InputAccessor> accessor)
+{
+    inputAccessors.emplace(accessor->number, accessor);
+}
+
+std::string EvalState::encodePath(const SourcePath & path)
+{
+    /* For backward compatibility, return paths in the root FS
+       normally. Encoding any other path is not very reproducible (due
+       to /nix/store/virtual000...<N>) and we should deprecate it
+       eventually. So print a warning about use of an encoded path in
+       decodePath(). */
+    return path.accessor == rootFS
+        ? path.path.abs()
+        : fmt("%s%08x-source%s", virtualPathMarker, path.accessor->number, path.path.absOrEmpty());
+}
+
+SourcePath EvalState::decodePath(std::string_view s, PosIdx pos)
+{
+    if (!hasPrefix(s, "/"))
+        throwEvalError(pos, "string '%1%' doesn't represent an absolute path", s);
+
+    if (hasPrefix(s, virtualPathMarker)) {
+        auto fail = [s]() {
+            throw Error("cannot decode virtual path '%s'", s);
+        };
+
+        s = s.substr(virtualPathMarker.size());
+
+        try {
+            auto slash = s.find('/');
+            size_t number = std::stoi(std::string(s.substr(0, slash)), nullptr, 16);
+            s = slash == s.npos ? "" : s.substr(slash);
+
+            auto accessor = inputAccessors.find(number);
+            if (accessor == inputAccessors.end()) fail();
+
+            SourcePath path {accessor->second, CanonPath(s)};
+
+            warn("applying 'toString' to path '%s' and then accessing it is deprecated, at %s", path, positions[pos]);
```

Comment:
Two problems related to https://github.com/NixOS/nix/pull/6530#discussion_r1029941664:
1. `toString` apparently doesn't return the virtual path if the resulting string is used for `builtins.{trace,throw}` ...?
2. It throws the deprecation warning for accessing the path when you do this, even though it's not accessed

This leads to this beautifully inconsistent example:
```
nix-repl> string = toString lazyPath

nix-repl> builtins.trace string string
warning: applying 'toString' to path '/home/tweagysil/src/nix/' and then accessing it is deprecated, at «stdin»:0
trace: /home/tweagysil/src/nix/
"/nix/store/virtual0000000000000000000000005-source"
```

Reply:
There is no inconsistency. The decoding of virtual paths is done by `builtins.{trace,throw}`, not by `toString`.

Reply:
Oh I see, but that's very problematic for debugging, because one would expect `builtins.{trace,throw}` to print the exact string that is being passed, not for it to do some processing. Here's a concrete example:

```
nix-repl> equals = a: b: builtins.trace "Checking whether string \"${a}\" is equal to string \"${b}\"" (a == b)                                                          

nix-repl> equals "/home/tweagysil/src/nix/" (toString lazyPath)                                                                                                          
warning: applying 'toString' to path '/home/tweagysil/src/nix/' and then accessing it is deprecated, at «stdin»:0
trace: Checking whether string "/home/tweagysil/src/nix/" is equal to string "/home/tweagysil/src/nix/"
false
```